projects/client/RabbitMQ.Client/src/client/impl/Frame.cs

```diff
@@ -1,218 +1,327 @@
+            
+            writer.Write((ushort)header.ProtocolClassId);
+            header.WriteTo(writer, (ulong)bodyLength);
+        }
+    }
+    
+    public class BodySegmentWriteFrame : WriteFrame
+    {
+        public BodySegmentWriteFrame(int channel, byte[] body, int offset, int count) : base(FrameType.FrameBody, channel)
+        {
+            NetworkBinaryWriter writer = base.GetWriter();
+            
+            writer.Write(body, offset, count);
+        }
+    }
+    
+    public class MethodWriteFrame : WriteFrame
+    {
+        public MethodWriteFrame(int channel, MethodBase method) : base(FrameType.FrameMethod, channel)
+        {
+            NetworkBinaryWriter writer = base.GetWriter();
+            
+            writer.Write((ushort)method.ProtocolClassId);
+            writer.Write((ushort)method.ProtocolMethodId);
+            
+            var argWriter = new MethodArgumentWriter(writer);
+            
+            method.WriteArgumentsTo(argWriter);
+            
+            argWriter.Flush();
+        }
+    }
+    
+    public class EmptyWriteFrame : WriteFrame
+    {
+        private static readonly byte[] m_emptyByteArray = new byte[0];
+
+        public EmptyWriteFrame() : base(FrameType.FrameHeartbeat, 0)
+        {
+            base.GetWriter().Write(m_emptyByteArray);
+        }
+        
+        public override string ToString()
+        {
+            return base.ToString() + string.Format("(type={0}, channel={1}, {2} bytes of payload)",
+                Type,
+                Channel,
+                Payload == null
+                    ? "(null)"
+                    : Payload.Length.ToString());
+        }
+    }
+    
+    public class WriteFrame : Frame
+    {
+        private readonly MemoryStream m_accumulator ;
+        private readonly NetworkBinaryWriter writer;
+        
+        public WriteFrame(FrameType type, int channel) : base(type, channel)
+        {
+            m_accumulator = new MemoryStream();
+            writer = new NetworkBinaryWriter(m_accumulator);
+        }
+        
+        public NetworkBinaryWriter GetWriter()
+        {
+            return writer;
+        }
+        
+        public override string ToString()
+        {
+            return base.ToString() + string.Format("(type={0}, channel={1}, {2} bytes of payload)",
+                Type,
+                Channel,
+                Payload == null
+                    ? "(null)"
+                    : Payload.Length.ToString());
+        }
+        
+        public void WriteTo(NetworkBinaryWriter writer)
+        {
+            var payload = m_accumulator.ToArray();
+            
+            writer.Write((byte)Type);
+            writer.Write((ushort)Channel);
+            writer.Write((uint)payload.Length);
+            writer.Write(payload);
+            writer.Write((byte)Constants.FrameEnd);
+        }
+    }
+    
+    public class ReadFrame : Frame

```

Comment:
Can we please rename this to `ReadableFrame` or `IncomingFrame`? There are no "write frames" in the protocol (that would be different from "read frames").